src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -0,0 +1,401 @@
+                _ids = new string[entryCount];
+                _lengths = new int[entryCount];
+
+                for (int i = 0; i < entryCount; ++i)
+                {
+                    LoadEntryAt(fs, indexOffset + (entrySize*i), out string id, out int byteOffset, out int length);
+
+                    _byteOffsets[i] = byteOffset + dataOffset;
+                    _ids[i] = id;
+                    _lengths[i] = length;
+
+                    if (length < 24) // Header Size
+                    {
+                        throw new InvalidOperationException(SR.InvalidOperation_BadIndexLength);
+                    }
+                }
+            }
+
+            private void ReadTzDataIntoBuffer(Stream fs, long position, Span<byte> buffer)
+            {
+                fs.Position = position;
+
+                int bytesRead = 0;
+                int bytesLeft = buffer.Length;
+
+                while (bytesLeft > 0)
+                {
+                    int b = fs.Read(buffer.Slice(bytesRead));
+                    if (b == 0)
+                    {
+                        break;
+                    }
+
+                    bytesRead += b;
+                    bytesLeft -= b;
+                }
+
+                if (bytesLeft != 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_ReadTZError, _tzFilePath, position, buffer.Length, bytesRead, buffer.Length));
+                }
+            }
+
+            private void LoadEntryAt(Stream fs, long position, out string id, out int byteOffset, out int length)
+            {
+                const int size = 52; // data entry size
+                Span<byte> entryBuffer = stackalloc byte[size];
+
+                ReadTzDataIntoBuffer(fs, position, entryBuffer);
+
+                int index = 0;
+                while (entryBuffer[index] != 0 && index < 40)
+                {
+                    index += 1;
+                }
+                id = Encoding.UTF8.GetString(entryBuffer.Slice(0, index));
+                byteOffset = TZif_ToInt32(entryBuffer.Slice(40, 4));
+                length = TZif_ToInt32(entryBuffer.Slice(44, 4));
+            }
+
+            public string[] GetTimeZoneIds()
+            {
+                return _ids;
+            }
+
+            public string GetTimeZoneDirectory()
+            {
+                return _tzFilePath;
+            }
+
+            public byte[] GetTimeZoneData(string id)
+            {
+                int i = Array.BinarySearch(_ids, id, StringComparer.Ordinal);
+                if (i < 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.TimeZoneNotFound_MissingData, id));
+                }
+
+                int offset = _byteOffsets[i];
+                int length = _lengths[i];
+
+                byte[] buffer = new byte[length];
+                ReadTzDataIntoBuffer(File.OpenRead(_tzFilePath), offset, buffer);
```

Comment:
Does this need `using` to close the file handle once it is done?

Reply:
Thanks, changed to 
```
using (FileStream fs = File.OpenRead(_tzFilePath))
{
    ReadTzDataIntoBuffer(fs, offset, buffer);
}
```